catalog/CatalogRelation.hpp

```diff
@@ -203,35 +242,83 @@ class CatalogRelation : public CatalogRelationSchema {
-    if (it != indices_.end()) {
-      return true;
-    } else {
+  bool hasIndexWithName(const std::string &index_name) const {
+    if (index_scheme_.get() == nullptr) {
       return false;
+    } else {
+      return index_scheme_->hasIndexWithName(index_name);
     }
   }
 
   /**
-   * @brief Create an index with a given name
+   * @brief Check whether an index with the given description
+   *        containing the same attribute id and index type
+   *        exists or not in the index map.
    *
-   * TODO (ssaurabh): This is only a placeholder implementation
-   * that just adds an index name into an indicies vector. This
-   * will be replaced by a full-fledged index data structure.
+   * @param index_descripton Index Description to check against.
+   * @return Whether a similar index description was already defined or not.
+   **/
+  bool hasIndexWithDescription(const IndexSubBlockDescription &index_description) const {
+    if (index_scheme_.get() == nullptr) {
+      return false;
+    } else {
+      return index_scheme_->hasIndexWithDescription(index_description);
+    }
+  }
+
+  /**
+   * @brief Add an index to the index_map.
    *
-   * @param The index to create
+   * @param index_name Name of the index to be added.
+   * @param index_description Corresponding description of the index.
+   * @return Whether the index was added successfully or not
    **/
-  void addIndex(const std::string& name) {
-    SpinSharedMutexExclusiveLock<false> lock(indices_mutex);
-    indices_.emplace_back(name);
+  bool addIndex(const std::string &index_name, const std::vector<IndexSubBlockDescription> &index_descriptions) {
+    // create an index_scheme, if it does not exist
+    if (index_scheme_.get() == nullptr) {
+      index_scheme_.reset(new IndexScheme());
+    }
+
+    // verify that index with the given name does not exist
+    if (index_scheme_->hasIndexWithName(index_name)) {
+      return false;
+    }
+
+    // verify that index with similar description does not exist
+    std::vector<IndexSubBlockDescription>::const_iterator itr;
+    for (itr = index_descriptions.begin(); itr != index_descriptions.end(); ++itr) {
+      if (index_scheme_->hasIndexWithDescription(*itr)) {
+        return false;
+      }
+    }
+
+    // acquire the lock and add the index
+    {
+      SpinSharedMutexExclusiveLock<false> lock(index_scheme_mutex);
+      {
+        SpinSharedMutexExclusiveLock<false> lock(layout_mutex_);
+        StorageBlockLayoutDescription *layout = default_layout_->getDescriptionMutable();
+        for (itr = index_descriptions.begin(); itr != index_descriptions.end(); ++itr) {
+          // add a new index
+          layout->add_index_description();
+          std::size_t num_indicies = layout->index_description_size();
+          // update the description of newly added index
+          // whose (position = num_indices - 1)
+          layout->mutable_index_description(num_indicies - 1)->CopyFrom(*itr);
```

Comment:
I think the following refactor should work:

```
layout->add_index_description()->MergeFrom(*itr);
```

The reason why we use `MergeFrom`, instead of `CopyFrom`, is for a better performance. We don't need to free up the space for the newly-added index description.